### ISSUE TITLE
GIF Error: Invalid colormap index when opacity index is outside of colormap

### ISSUE THREAD

**Issue by dloebl** [ImageMagick/ImageMagick#4634]
### ImageMagick version

7.1.0-18

### Operating system

Linux

### Operating system, version and so on

ArchLinux

### Description

When the opacity index of a GIF frame is outside of the colormap ImageMagick fails with the following error:
`convert: invalid colormap index input.gif @ error/colormap-private.h/ConstrainColormapIndex/35.`
It is valid to set the opacity index outside of the colormap (for an GIF encoder), as the RGB component is not needed.

**For instance:**
`opacity index = 255`
`size of colormap = 128`

Attached are two samples.

### Steps to Reproduce

`$ convert input.gif out.gif`

### Images

![input](https://user-images.githubusercontent.com/34104349/147479738-adfbe1e6-22ea-44c0-954a-ef1c6bb0c7df.gif)
![case2](https://user-images.githubusercontent.com/34104349/147480317-98fc781c-1bdc-4090-afad-4fc8d7976d43.gif)

**Comment by dlemstra** [comment]
Why do you think that it is valid to set the index outside of the colormap?

**Comment by urban-warrior** [comment]
Transparency in GIF is enabled by setting the Transparency Color flag to 1.  The transparent color must be in the active color table.  For the striped image, there is a color table of 4 colors with a transparent index of 7, thus ImageMagick throws an exception.  ImageMagick is security focused and therefore not tolerant of garbage data in an image format.

If you can point to verbiage in the GIF specification that permits a transparent color index to be outside the active color table, we certainly will consider merging your pull request.

**Comment by dloebl** [comment]
**Maybe I should give a few more details:** 
Say one has 128 colors in the color palette (all colors in use) and uses index 128 as the transparency index on top. In this case, the initial LZW code length has to be increased because the index used for transparency must be at least 128 and so needs one bit more to represent it than 0-127. At the moment, ImageMagick increases the number of colors ([coders/gif.c#L1278](https://github.com/ImageMagick/ImageMagick/blob/main/coders/gif.c#L1278)) to 129 in such a case and accepts 128 as the transparent index. The [GIF standard](https://www.w3.org/Graphics/GIF/spec-gif89a.txt) requires that the number of entries in the color table is always a power of two. So if index 128 is accepted as a valid index, then it would make sense to accept also all the other numbers below the next power of two as the transparent index (e.g. 129-255). 
Regarding the security check, I suggest that you throw the "invalid colormap index" error if an index is used that is outside the color table and also is not the transparent index (done in my PR).

**Comment by urban-warrior** [comment]
If you can point to verbiage in the GIF specification that permits a transparent color index to be outside the active color table, we certainly will consider merging your pull request.

**Comment by dloebl** [comment]
Hi again,

Thanks for your fast reply.

Strictly speaking, setting the transparency index outside of the color table is not covered by the standard:
[**22. Table Based Image Data:**](https://www.w3.org/Graphics/GIF/spec-gif89a.txt) `[..] Each index must be within the range of the size of the active color table, starting at 0.`

However, as pointed out in [my last comment](https://github.com/ImageMagick/ImageMagick/issues/4634#issuecomment-1002753304), ImageMagick already allows the transparency index to be exactly one position outside of the color table ([ImageMagick/blob/main/coders/gif.c#L1278](https://github.com/ImageMagick/ImageMagick/blob/main/coders/gif.c#L1278)).

For some GIFs, it also makes somehow sense to set the transparency index outside of the active color table.
In the case of e.g. 128 colors plus transparency index, the [GIF standard](https://www.w3.org/Graphics/GIF/spec-gif89a.txt) would otherwise require the following: The color table must be filled up with another 128 dummy colors because GIF just allows powers of two for the color table size (see: [19. Global Color Table](https://www.w3.org/Graphics/GIF/spec-gif89a.txt), [21. Local Color Table](https://www.w3.org/Graphics/GIF/spec-gif89a.txt)). From a practical perspective, this approach does not make so much sense since none of these new color indices are ever addressed and so space is wasted. To save space, GIFs with the transparency index outside of the color table exist, and I would guess that line [gif.c:1278](https://github.com/ImageMagick/ImageMagick/blob/main/coders/gif.c#L1278) has been included for this reason.

**To sum it up:**
ImageMagick currently allows one specific transparency index to be outside the range of the active color table. If one specific index outside of the active color table is already allowed, it would make sense to allow any index which can be represented with the LZW code size (see: [Appendix F. Variable-Length-Code LZW Compression](https://www.w3.org/Graphics/GIF/spec-gif89a.txt)).

I hope this explanation is helpful. Please let me know if there are changes required on my pull request.

**Comment by urban-warrior** [comment]
> ImageMagick already allows the transparency index to be exactly one position outside of the color table

The color table is increased by one entry and the transparency color is initialized.  When the GIF image is read, the transparency index defined in the graphics control extension is within the range of the size of the active color table.

**Comment by MCLoebl** [comment]
Hi, @dloebl asked me to have a look into this issue. My impression is that there is a bit of a misunderstanding here. On the one hand, ImageMagick internally increases the number of colors by +1 when transparency is used. On the other hand, @dloebl is referring to the color table in the GIF file itself (which is always a power of two in size).

The ImageMagick decoder accepts GIFs where the transparency index is outside the GIF's color table by exactly +1. All other transparency indices below the next power of two are not accepted. You find a simple illustration below the text. The point is that both kinds of GIFs ((a) transparency index too large by +1, (b) transparency index too large by +2,3...) are strictly speaking against the GIF89a standard. One can argue that there is an inconsistency since the ImageMagick GIF-decoder accepts case (a), but not case (b). In the PR, @dloebl suggests accepting both cases (a) and (b) as most GIF-decoders do. The alternative would be to reject both.

I hope that this explanation is helpful. I would say that this issue is not about a bug in ImageMagick, but maybe more about a minor inconsistency. Please, let me know if you have any questions.

<img width="556" alt="snapshot" src="https://user-images.githubusercontent.com/79047536/148080517-38a29ae6-53b6-4e24-ad5e-a645c234b716.png">

**Comment by dlemstra** [comment]
It seems that we added this as a bugfix to resolve the issue reported in #2007. This was added to allow an off by one error for the transparency index. Not sure if we should at this point in time revert that patch but also not sure if we should extend the pallete by the power of two for this other situation.
